rootfs/etc/nginx/lua/balancer/sticky.lua

```diff
@@ -59,31 +67,82 @@ local function set_cookie(self, value)
   end
 end
 
+function _M.get_last_failure()
+  return ngx_balancer.get_last_failure()
+end
+
 function _M.balance(self)
   local cookie, err = ck:new()
   if not cookie then
     ngx.log(ngx.ERR, "error while initializing cookie: " .. tostring(err))
     return
   end
 
+  -- upstream_from_cookie: upstream which is pointed by sticky cookie
+  local upstream_from_cookie = nil
+
   local key = cookie:get(self:cookie_name())
-  if not key then
+  if key then
+    upstream_from_cookie = self.instance:find(key)
+  end
+
+  -- get state of the previous attempt
+  local state_name = self.get_last_failure()
+
+  -- use previous upstream if this is the first attempt or previous attempt succeeded
+  if state_name == nil and upstream_from_cookie ~= nil then
+    do return upstream_from_cookie end
```

Comment:
why `do ... end` ?

Reply:
Yeah, right. They are not required.  
_removed in ddf738fddadfb7ea9090ed2bd6fe277a34dcb81e_